src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs

```diff
@@ -1265,106 +1280,133 @@ protected override void ProcessRecord()
-                                    response.Content.Headers.ContentLength,
-                                    contentType);
-                                WriteVerbose(respVerboseMsg);
-
-                                if (!response.IsSuccessStatusCode)
+                                FillRequestStream(request);
+                                try
                                 {
-                                    string message = String.Format(CultureInfo.CurrentCulture, WebCmdletStrings.ResponseStatusCodeFailure,
-                                        (int)response.StatusCode, response.ReasonPhrase);
-                                    HttpResponseException httpEx = new HttpResponseException(message, response);
-                                    ErrorRecord er = new ErrorRecord(httpEx, "WebCmdletWebResponseException", ErrorCategory.InvalidOperation, request);
-                                    string detailMsg = "";
-                                    StreamReader reader = null;
-                                    try
+                                    long requestContentLength = 0;
+                                    if (request.Content != null)
+                                        requestContentLength = request.Content.Headers.ContentLength.Value;
+
+                                    string reqVerboseMsg = String.Format(CultureInfo.CurrentCulture,
+                                        WebCmdletStrings.WebMethodInvocationVerboseMsg,
+                                        request.Method,
+                                        request.RequestUri,
+                                        requestContentLength);
+                                    WriteVerbose(reqVerboseMsg);
+                                    
+                                    response = GetResponse(client, request, stripAuthorization);
+                                    
+                                    bool causeFailureForTesting = (System.Management.Automation.Internal.InternalTestHooks.FailWebRequestCount > 0) && retriesLeft > 0;
+
+                                    isSuccessful = response.IsSuccessStatusCode && !causeFailureForTesting;                                
+
+                                    if (!isSuccessful && retriesLeft > 0)
+                                    {
+                                        retriesLeft--;
+
+                                        string retryMessage = string.Format(CultureInfo.CurrentCulture,
+                                            WebCmdletStrings.RetryVerboseMsg,
+                                            RetryIntervalSec,
+                                            response.StatusCode
+                                            );
+
+                                        WriteVerbose(retryMessage);
+                                        Task.Delay(RetryIntervalSec * 1000).GetAwaiter().GetResult();
+                                        continue;
```

Comment:
We should probably dispose the current HttpResponseMessage before retrying.

Reply:
fixed